File: typescript/src/model/template.ts

///class for the template property of RalObjects and RalMethods
export class Template {
    constructor(
        public ralType: string,
        public version: string,
        public objectStateTemplates: string | null = null,
    ) { }

    toMap(): Record<string, any> {
        return {
            "RALType": this.ralType,
            "version": this.version,
            "objectStateTemplates": this.objectStateTemplates ?? null,
        };
    }

    static fromMap(map: Record<string, any>): Template {
        const { RALType, version, objectStateTemplates } = map;

        //check that RALType and version are defined
        if (RALType == null) {
            throw new Error('RALType is not defined in the map');
        }
        if (version == null) {
            throw new Error('version is not defined in the map');
        }

        return new Template(
            RALType,
            version,
            objectStateTemplates ?? ""
        );
    }
}

File: dart/lib/src/model/template.dart
import 'package:either_dart/either.dart';
import 'package:openral_core/src/cross/backend/parsing/parsing_error.dart';
import 'package:openral_core/src/model_parser/template_parser.dart';

class Template {
  final String ralType;
  final String version;
  final List<String> objectStateTemplates;

  Template({
    required this.ralType,
    required this.version,
    this.objectStateTemplates = const [],
  });

  Map<String, dynamic> toMap() {
    return {
      'RALType': ralType,
      'version': version,
      'objectStateTemplates': objectStateTemplates,
    };
  }

  static Either<Template, ParsingError> fromMap(map) {
    final templateParser = TemplateParser();

    return templateParser.fromMap(map);
  }
}
